machine: add Microblaze support for Zephyr toolchain (32-bit, little-endian)

alpsayin opened the pull request:
Utilises the existing sources. Forked off of `zephyr`'s `picolibc` mirror to be able to collect consistent results from its CI.

Helps meson pick the `microblaze` cpu family from `microblazeel` tool.
`do-microblazeel-configure` then picks `cross-microblazeel-zephyr-elf.txt`
`zephyr` only supports `little-endian` so far _(last I checked)_, thus I never bothered to make Zephyr `sdk-ng` produce a big-endian toolchain. Hence no picolibc support for `microblaze-zephyr-elf` (big-endian is default).

No QEMU test set because QEMU invokation would need a `-hw-dtb` file such as 
https://github.com/zephyrproject-rtos/zephyr/blob/713b955801c44aadf2e66ed972b83da8ca03d837/boards/microblaze/qemu_microblaze/board-qemu-microblaze-demo.dtb

But @keith-packard and I agree that, that's the next step. Let's see with this PR first if it builds.

Future work?
1. Vitis shipped `microblaze-xilinx-elf` toolchain support?

EDIT/Update: it seems there is no CI here.
EDIT/Update2: Of course there isn't. I'll open a draft PR in `sdk-ng` as usual.
EDIT/Update3: Future work.

keith-packard commented on .github/do-linux:
```diff
@@ -29,6 +29,7 @@ HERE=`dirname "$0"`
 "$HERE"/do-test clang-rv32imafdc "$@"
 "$HERE"/do-zephyr-build arc "$@"
 "$HERE"/do-zephyr-build arc64 "$@"
+"$HERE"/do-zephyr-build microblaze "$@"
```
These builds rely on Zephyr SDK 0.15.1, so I don't think we can include this bit yet.

alpsayin replied:
Gone.

keith-packard commented on README.md:
```diff
@@ -41,6 +41,7 @@ at this point only has code to build for the following targets:
  * ARM (32- and 64- bit)
  * i386 (Native and Linux hosted, for testing)
  * Motorola 68000 (m68k)
+ * Microblaze (32-bit)
```
Should probably mention that this is for both big- and little- endian targets.

alpsayin replied:
Fixed.

keith-packard commented on newlib/libc/machine/microblaze/CMakeLists.txt:
```diff
@@ -0,0 +1,48 @@
+#
+# SPDX-License-Identifier: BSD-3-Clause
+#
+# Copyright © 2022 Keith Packard
```
This should carry just your copyright (and your employer if necessary)

alpsayin replied:
Done.

keith-packard commented on newlib/libc/machine/microblaze/meson.build:
```diff
@@ -0,0 +1,47 @@
+#
+# SPDX-License-Identifier: BSD-3-Clause
+#
+# Copyright © 2022 Keith Packard
```
Same as CMakeLists.txt

alpsayin replied:
Fixed.

keith-packard commented on scripts/do-microblaze-configure:
```diff
@@ -0,0 +1,38 @@
+#!/bin/sh
```
should be named do-microblazeel-configure to allow for a future big-endian version

alpsayin replied:
Done.

keith-packard commented on scripts/do-microblaze-configure:
```diff
@@ -0,0 +1,38 @@
+#
+# SPDX-License-Identifier: BSD-3-Clause
+#
+# Copyright © 2019 Keith Packard
```
Same copyright issue as above

alpsayin replied:
Fixed. and one other.

keith-packard reviewed (changes requested):
How much testing has this seen? Did you actually try the machine-specific APIs on a target?

Only changes required are to not include this in CI (yet), and fix the copyrights to exclude me from the list.

keith-packard commented:
When you've fixed the CI issues, I can let CI run, but it won't tell us much. Please let me know how much testing you've managed to do.

alpsayin commented:
> How much testing has this seen? Did you actually try the machine-specific APIs on a target?

1st iteration: None.
Current [iteration b52120d](https://github.com/picolibc/picolibc/commit/b52120dceaf94c8b84cf8d8333d441647d5ee7ad): It builds and builds tests if I disable `atomic-ungetc`. I was also able to compile a no-stdio C program with the output specs.
Using the Zephyr SDK toolchain `zephyr-sdk-0.16.0-beta1-7-g2a8106a` from  https://github.com/zephyrproject-rtos/sdk-ng/pull/630
I'll try to load some of the test executables into a QEMU machine.
I may also write a run-microblaze script to automate testing the executables.

alpsayin commented:
> Did you actually try the machine-specific APIs on a target?

Could you elaborate on what you mean by machine-specific APIs?

keith-packard commented:
> > How much testing has this seen? Did you actually try the machine-specific APIs on a target?
> 
> 1st iteration: None. Current [iteration b52120d](https://github.com/picolibc/picolibc/commit/b52120dceaf94c8b84cf8d8333d441647d5ee7ad): It builds and builds tests if I disable `atomic-ungetc`.

Hrm. What's the deal with atomic ungetc? Does the target have atomic operations of some size? Is it like RISC-V and only support atomic 32- and 64- bit operations? (there are code paths for that, btw). Without atomic ungetc, stdio isn't re-entrant as it doesn't otherwise need (or provide) any locking.

> I was also able to compile a no-stdio C program with the output specs. Using the Zephyr SDK toolchain `zephyr-sdk-0.16.0-beta1-7-g2a8106a` from [zephyrproject-rtos/sdk-ng#630](https://github.com/zephyrproject-rtos/sdk-ng/pull/630) I'll try to load some of the test executables into a QEMU machine. I may also write a run-microblaze script to automate testing the executables.

That's awesome. If you manage that, and can do something like the arm semi-hosting (or even the x86 'e9' hack) you should be able to run the complete picolibc test suite without any code outside of picolibc. There's no requirement for merging to picolibc though; lots of hosts just don't have any semihosting support yet.

keith-packard commented:
> > Did you actually try the machine-specific APIs on a target?
> 
> Could you elaborate on what you mean by machine-specific APIs?

Yes: all of the code in newlib/libc/machine/microblaze. None of that will get any testing using other targets, so some level of validation would be helpful, even if done just once by hand.

keith-packard commented on newlib/libc/machine/microblaze/strcpy.c:
```diff
@@ -90,7 +90,7 @@ strcpy (char *__restrict dst0,
   char *s = dst0;
 
-  while (*dst0++ = *src0++)
+  while ((*dst0++ = *src0++))
```
thanks for the cleanup

alpsayin replied:
14k files being compiled. 5k x 3 lines of warnings of `hey you should put that into a parentheses` :D
Another one was `-fstack-protector is not supported` warning. Took care of it too.

keith-packard commented on scripts/cross-microblazeel-zephyr-elf.txt:
```diff
@@ -0,0 +1,20 @@
+# so we have to add -nostdlib to the compiler configuration itself or
+# early compiler tests will fail. This can be removed when picolibc
+# requires at least version 0.54.2 of meson.
+c = ['microblazeel-zephyr-elf-gcc', '-nostdlib', '-mxl-soft-mul', '-mxl-soft-div', '-msoft-float']
```
are these compiler flags for math operations not provided in a multilib configuration?

alpsayin replied:
Seems it does. I removed those options. I have very little trust for microblaze-gcc...

```
microblazeel-zephyr-elf-gcc --print-multi-lib
.;
bs;@mxl-barrel-shift
m;@mno-xl-soft-mul
le;@mlittle-endian
m/mh;@mno-xl-soft-mul@mxl-multiply-high
m/le;@mno-xl-soft-mul@mlittle-endian
m/mh/le;@mno-xl-soft-mul@mxl-multiply-high@mlittle-endian
bs/m;@mxl-barrel-shift@mno-xl-soft-mul
bs/le;@mxl-barrel-shift@mlittle-endian
bs/m/mh;@mxl-barrel-shift@mno-xl-soft-mul@mxl-multiply-high
bs/m/le;@mxl-barrel-shift@mno-xl-soft-mul@mlittle-endian
bs/m/mh/le;@mxl-barrel-shift@mno-xl-soft-mul@mxl-multiply-high@mlittle-endian
```

keith-packard reviewed (changes requested):
The spurious #endif in abort.c seems wrong to me?

keith-packard commented on newlib/libc/machine/microblaze/abort.c:
```diff
@@ -88,3 +88,4 @@ abort (void)
       exit(1);
     }
 }
+#endif
```
Ah, someone was using a compiler that didn't like files which generated no symbols. Good catch, and sorry for the false alarm :-)

keith-packard commented:
We'll let CI run on this now; when that passes (which it should, unless we've missed something), I'll get it merged.

alpsayin commented:
1. Rebased on `main`
2. Renamed `scripts/do-microblaze-configure` -> `scripts/do-microblazeel-configure`
3. Disabled `fstack-protector` for tests to stop build from moaning about `unsupported option`
4. while loop condition that is also an assignment is placed in parentheses to stop build warnings.
5. Fixed licenses.
